File: VIF_git/GSDS-Qual/data-structure/binary_tree/binary_tree_array.py
"""
implementation of bianary_tree with array

indexing
    find child given parent; parent index: n
        left child: (2 * n + 1)
        right child: (2 * n + 2)

    find parent given child; child index: n
        parent index = (n - 1) // 2

** sanity check 필요
    현재 인덱스가 유효한 인덱스인지 체크
"""

def traversal_from_root(tree):
    i = 0
    n = len(tree)
    while i < n:
        if tree[i]:
            print(f"Parent: {tree[i]}", end = ", ")
            left = 2 * i + 1
            right = left + 1
            if left < n and tree[left] is not None:
                print(f"Left: {tree[left]}", end = ", ")
            if right < n and tree[right] is not None:
                print(f"Right: {tree[right]}", end = " ")
            print()
        i += 1

def traversal_from_leaf(tree):
    n = len(tree)
    i = n-1
    while i > 0:
        if tree[i]:
            print(f"Parent of {tree[i]} -> {tree[(i-1)//2]}")
        i -= 1


if __name__ == "__main__":
    tree = ["A", "B", "C", "D", "E", "F", None, "G"]
    print("Finding children given parents.")
    traversal_from_root(tree)
    print()
    print("Finding parent given child.")
    traversal_from_leaf(tree)

File: VIF_git/GSDS-Qual/data-structure/binary_tree/binary_tree_array.c
#include <stdio.h>
#include <string.h>

#define MAX_SIZE 100

// 배열 기반 이진 트리 순회 - 부모로부터 자식 노드 찾기
void traversal_from_root(char tree[], int size) {
    int i = 0;
    while (i < size) {
        if (tree[i] != '\0') { // '\0'은 빈 값을 나타냅니다.
            printf("Parent: %c", tree[i]);

            int left = 2 * i + 1;
            int right = left + 1;

            if (left < size && tree[left] != '\0') {
                printf(", Left: %c", tree[left]);
            }
            if (right < size && tree[right] != '\0') {
                printf(", Right: %c", tree[right]);
            }
            printf("\n");
        }
        i++;
    }
}

// 배열 기반 이진 트리 순회 - 자식으로부터 부모 노드 찾기
void traversal_from_leaf(char tree[], int size) {
    int i = size - 1;
    while (i > 0) {
        if (tree[i] != '\0') {
            printf("Parent of %c -> %c\n", tree[i], tree[(i - 1) / 2]);
        }
        i--;
    }
}

int main() {
    // 트리 배열 초기화
    char tree[MAX_SIZE];
    memset(tree, '\0', sizeof(tree)); // 모든 요소를 '\0'으로 초기화

    // 트리 데이터를 삽입
    tree[0] = 'A';
    tree[1] = 'B';
    tree[2] = 'C';
    tree[3] = 'D';
    tree[4] = 'E';
    tree[5] = 'F';
    tree[7] = 'G';

    // 부모에서 자식 노드 찾기
    printf("Finding children given parents.\n");
    traversal_from_root(tree, MAX_SIZE);
    printf("\n");

    // 자식에서 부모 노드 찾기
    printf("Finding parent given child.\n");
    traversal_from_leaf(tree, MAX_SIZE);

    return 0;
}